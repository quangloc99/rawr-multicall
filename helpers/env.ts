import * as z from 'zod';
import { config } from 'dotenv';

config();

export const ENV_SCHEMA = z.object({
    BLOCK_NUMBER: z.coerce.number().optional(),
    CHAIN_ID: z.coerce.number(),
    ALL_CHAIN: z.coerce.boolean().optional(),
});

export type Env = z.infer<typeof ENV_SCHEMA>;
export const env = ENV_SCHEMA.parse(process.env);
