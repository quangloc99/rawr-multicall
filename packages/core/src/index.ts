export * from './Address';
export * from './InstructionContext';
export * from './instructions';
export * from './Bytes';
export * from './Call';
export * from './buildContract';
export * from './buildRawMulticallContract';
export * from './decodeOutput';
export * from './errors';
export * from './registerPredeployContract';
export * from './Result';
