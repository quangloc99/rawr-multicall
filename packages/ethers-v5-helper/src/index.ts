export * from './types';
export * from './createEthersV5Call';
export * from './error';
