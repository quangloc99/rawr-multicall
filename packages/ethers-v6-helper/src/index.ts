export * from './types';
export * from './error';
export * from './createEthersV6Call';
