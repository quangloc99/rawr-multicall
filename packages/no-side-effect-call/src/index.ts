import { registerPredeployCallThenRevert } from './registerPredeployCallThenRevert';
registerPredeployCallThenRevert();

export * from './CallThenRevertContractArtifact';
export * from './registerPredeployCallThenRevert';
export * from './wrapNoSideEffectCall';
