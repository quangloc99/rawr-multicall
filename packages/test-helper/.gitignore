artifacts
ethers-v6-contracts/typechain-types