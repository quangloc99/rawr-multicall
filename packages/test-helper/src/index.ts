import { CHAIN_ID_MAPPING } from './chainId';
import { testEnv } from './test-env';
import { RPC_URL } from './rpc-url';

export { testEnv, RPC_URL, CHAIN_ID_MAPPING };
export * from './jest.config';
export * as testData from './test-data';

export function describeForChain(
    ...params:
        | [callback: (rpcUrl: string, chain: number) => void]
        | [nameFormatter: (chain: number) => string, callback: (rpcUrl: string, chain: number) => void]
) {
    const [nameFormatter, callback] =
        params.length == 2 ? params : [(chain: number) => `Testing chain ${chain}`, ...params];

    for (const chain of Object.values(CHAIN_ID_MAPPING)) {
        const isAllowedChain = testEnv.CHAIN_ID == undefined || chain === testEnv.CHAIN_ID;
        const name = nameFormatter(chain);
        (isAllowedChain ? describe : describe.skip)(name, () => callback(RPC_URL[chain], chain));
    }
}
