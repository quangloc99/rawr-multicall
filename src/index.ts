export * from './Address';
export * from './InstructionContext';
export * from './instructions';
export * from './bytes';
export * from './Call';
export * from './buildContract';
export * from './buildRawMulticallContract';
export * from './decodeOutput';
export * from './chainId';
export * from './errors';
