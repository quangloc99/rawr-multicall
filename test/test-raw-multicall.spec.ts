import { Bytes, buildRawMulticallContract, createCall } from '../src';
import { Interface, JsonRpcProvider } from 'ethers';
import { ERC20Abi } from './abi';
import { describeForChain } from './test-helper';
import { env } from '../helpers/env';

describeForChain(
    (chain: number) => `buildRawMulticallContract for chain ${chain}`,
    (rpcUrl: string) => {
        let provider: JsonRpcProvider;
        const blockTag = env.BLOCK_NUMBER ?? 17657895;

        const doSend = (data: Bytes) => provider.call({ data, blockTag });

        beforeAll(() => {
            provider = new JsonRpcProvider(rpcUrl);
        });

        describe('erc20', () => {
            const iface = new Interface(ERC20Abi);
            const usdc = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48';
            const pendle = '0x808507121b80c02388fad14726482e061b8da827';

            const holder = ['0xa3a7b6f88361f48403514059f1f16c8e78d60eec', '0x28c6c06298d514db089934071355e5743bf21d60'];

            it('simple', async () => {
                const callData = iface.encodeFunctionData('balanceOf', [holder[0]]);
                const sendData = buildRawMulticallContract([createCall(pendle, callData)]);
                expect(callData).toMatchSnapshot();
                expect(sendData).toMatchSnapshot();
                const res = await doSend(sendData.byteCode);
                expect(res).toMatchSnapshot();
            });
        });
    }
);
