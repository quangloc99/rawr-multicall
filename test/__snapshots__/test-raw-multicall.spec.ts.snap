// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildRawMulticallContract erc20 simple 1`] = `"0x70a08231000000000000000000000000a3a7b6f88361f48403514059f1f16c8e78d60eec"`;

exports[`buildRawMulticallContract erc20 simple 2`] = `
{
  "byteCode": "0x602462000046608039600460a45f5f602460805f73808507121b80c02388fad14726482e061b8da8275af181536001013d60e01b815281013d5f823e3d0160a4900360a4f30070a08231000000000000000000000000a3a7b6f88361f48403514059f1f16c8e78d60eec",
  "splittedByteCodes": [
    "6024",
    "62000046",
    "6080",
    "39",
    "6004",
    "60a4",
    "5f",
    "5f",
    "6024",
    "6080",
    "5f",
    "73808507121b80c02388fad14726482e061b8da827",
    "5a",
    "f1",
    "81",
    "53",
    "6001",
    "01",
    "3d",
    "60e0",
    "1b",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "60a4",
    "90",
    "03",
    "60a4",
    "f3",
    "00",
    "",
    "70a08231000000000000000000000000a3a7b6f88361f48403514059f1f16c8e78d60eec",
  ],
}
`;

exports[`buildRawMulticallContract erc20 simple 3`] = `"0x0100000020000000000000000000000000000000000000000000110dad9ac2ff4020aa7336"`;
